bc_obps/service/report_version_service.py

```diff
@@ -0,0 +1,67 @@
+from django.db import transaction
+from reporting.models.facility_report import FacilityReport
+from reporting.models.report import Report
+from reporting.models.report_operation import ReportOperation
+from reporting.models.report_version import ReportVersion
+from service.data_access_service.facility_service import FacilityDataAccessService
+
+
+class ReportVersionService:
+    @staticmethod
+    def create_report_version(
+        report: Report,
+        report_type: str = "Annual Report",
+    ) -> ReportVersion:
+        # Creating first version
+        report_version = ReportVersion.objects.create(report=report, report_type=report_type)
+        # Pre-populating data to the first version
+        operation = report.operation
+        operator = report.operator
+
+        report_operation = ReportOperation.objects.create(
+            operator_legal_name=operator.legal_name,
+            operator_trade_name=operator.trade_name,
+            operation_name=operation.name,
+            operation_type=operation.type,
+            operation_bcghgid=operation.bcghg_id.id if operation.bcghg_id else None,
+            bc_obps_regulated_operation_id=(
+                operation.bc_obps_regulated_operation.id if operation.bc_obps_regulated_operation else ""
+            ),
+            operation_representative_name=(
+                operation.point_of_contact.get_full_name() if operation.point_of_contact else ""
+            ),
+            report_version=report_version,
+        )
+        report_operation.activities.add(*list(operation.activities.all()))
+        report_operation.regulated_products.add(*list(operation.regulated_products.all()))
+
+        facilities = FacilityDataAccessService.get_current_facilities_by_operation(operation)
+
+        for f in facilities:
+            facility_report = FacilityReport.objects.create(
+                facility=f,
+                facility_name=f.name,
+                facility_type=f.type,
+                facility_bcghgid=f.bcghg_id.id if f.bcghg_id else None,
+                report_version=report_version,
+            )
+            facility_report.activities.add(*list(operation.activities.all()))
+
+        return report_version
+
+    @staticmethod
+    def delete_report_version(report_version_id: int) -> None:
+        report_version = ReportVersion.objects.get(id=report_version_id)
+        report_version.delete()
+
+    @staticmethod
+    @transaction.atomic
+    def change_report_version_type(report_version_id: int, new_report_type: str) -> ReportVersion:
+        report_version = ReportVersion.objects.get(id=report_version_id)
+        if report_version.report_type == new_report_type:
+            return report_version
+
+        ReportVersionService.delete_report_version(report_version.id)
```

Comment:
I filled out a report from beginning to end and then tried to change the report type. It errored out on protected foreign keys.
Looks like we have some incorrectly set `on_delete=PROTECT ` values set on a few of our models:
ReportAddionalData
ReportVerification
ReportProductEmissionAllocation
all have on_delete=PROTECT on their report_version fkey. We'll have to change those for this delete service to work on a fully complete report.